code/game/objects/structures/barsign.dm

```diff
@@ -41,363 +53,717 @@
+			. += "<span class='notice'>The frame needs <i>wiring</i> and the circuit board could be <b>pried out</b>.</span>"
+		if(BARSIGN_WIRED)
+			. += "<span class='notice'>The frame lacks a <i>glass screen</i> and is filled with wires that could be <b>cut</b>.</span>"
+		if(BARSIGN_COMPLETE)
+			. += "<span class='info'><b>Alt-Click</b> to toggle it's power.</span>"
+			if(panel_open)
+				. += "<span class='notice'>It is disabled by it's <i>unscrewed</i> maintenance panel that exposes an area from which the screen could be <b>pried out</b>.</span>"
+
+/obj/machinery/barsign/proc/is_on()
+	if(power_state == ACTIVE_POWER_USE)
+		return TRUE
+	return FALSE
+
+/obj/machinery/barsign/proc/toggle_on_off()
+	if(is_on())
+		return turn_off()
+	else
+		return turn_on()
+
+/obj/machinery/barsign/proc/turn_off()
+	if((stat & (BROKEN|EMPED|MAINT)) || !is_on() || build_stage < BARSIGN_COMPLETE)
+		return FALSE
+	save_sign(current_sign)
+	set_light(0)
+	power_state = IDLE_POWER_USE
+	set_sign(new /datum/barsign/hiddensigns/signoff)
+	playsound(src, 'sound/machines/lightswitch.ogg', 10, TRUE)
+	return TRUE
 
-/obj/structure/sign/barsign/attack_ai(mob/user as mob)
-	return src.attack_hand(user)
+/obj/machinery/barsign/proc/turn_on()
+	if((stat & (BROKEN|NOPOWER|MAINT)) || is_on() || build_stage < BARSIGN_COMPLETE)
+		return FALSE
+	if(panel_open)
+		return FALSE
+	set_light(1, LIGHTING_MINIMUM_POWER)
+	power_state = ACTIVE_POWER_USE
+	if(prev_sign)
+		set_sign(prev_sign)
+	else
+		set_random_sign()
+	playsound(src, 'sound/machines/lightswitch.ogg', 10, TRUE)
+	return TRUE
 
-/obj/structure/sign/barsign/attack_hand(mob/user as mob)
-	if(!src.allowed(user))
-		to_chat(user, "<span class = 'info'>Access denied.</span>")
+/obj/machinery/barsign/attack_hand(mob/user)
+	if(..())
 		return
-	if(broken)
-		to_chat(user, "<span class ='danger'>The controls seem unresponsive.</span>")
+	if(stat & MAINT)
+		to_chat(user, "<span class='warning'>Wait until the repairs are complete!</span>")
+		return
+	if((stat & (BROKEN|NOPOWER|EMPED)) || build_stage < BARSIGN_COMPLETE)
+		to_chat(user, "<span class='warning'>The controls seem unresponsive.</span>")
+		return
+	if(panel_open)
+		to_chat(user, "<span class='warning'>Close the maintenance panel first!</span>")
+		return
+	if(!allowed(user))
+		to_chat(user, "<span class='warning'>Access denied.</span>")
 		return
 	pick_sign()
 
-/obj/structure/sign/barsign/attackby(obj/item/I, mob/user)
-	if(istype(I, /obj/item/stack/cable_coil) && panel_open)
-		var/obj/item/stack/cable_coil/C = I
-		if(emagged) //Emagged, not broken by EMP
-			to_chat(user, "<span class='warning'>Sign has been damaged beyond repair!</span>")
+/obj/machinery/barsign/AltClick(mob/user)
+	add_fingerprint(user)
+	if(toggle_on_off())
+		return
+	attack_hand(user)
+
+/obj/machinery/barsign/attackby(obj/item/I, mob/user, params)
+	// Inserting the electronics/circuit
+	if(build_stage == BARSIGN_FRAME && istype(I, /obj/item/barsign_electronics))
```

Comment:
A switch could be used here?

Reply:
it uses && so I don't think it can, unless DM does something horrible with switch statements

Reply:
You could check for the type after the build_stage, returning early ..() if false, then it would be possible.